Added IMU support

nindanaoto opened the pull request:
Hello maintainers,

I added IMU support based on depthai_examples/src/stereo_publisher.cpp .
This PR needs https://github.com/luxonis/depthai-ros/pull/39 , so this PR should be merged after the dependent PR has been merged.

nindanaoto committed:
Added IMU support

saching13 reviewed (changes requested):
Thanks alot for the PR. 
If you are up for it. can you add depth align argument to this example along with rgb camera publishing when it is aligned to rgb camera ?

saching13 commented:
https://docs.luxonis.com/projects/api/en/latest/components/nodes/imu/#usage

imu->setMaxBatchReports(1);
Can you also set this. that way we will get 1 message only for now.

nindanaoto committed:
Limit IMU BatchReports to 1

nindanaoto commented:
@saching13 
> Thanks alot for the PR. If you are up for it. can you add depth align argument to this example along with rgb camera publishing when it is aligned to rgb camera ?

I will add the RGB camera publishing and a `depth_aligned` option based on this depthai-core example.
https://github.com/luxonis/depthai-core/blob/main/examples/StereoDepth/rgb_depth_aligned.cpp

saching13 commented:
> @saching13 
> 
> > Thanks alot for the PR. If you are up for it. can you add depth align argument to this example along with rgb camera publishing when it is aligned to rgb camera ?
> 
> 
> 
> I will add the RGB camera publishing and a `depth_aligned` option based on this depthai-core example.
> 
> https://github.com/luxonis/depthai-core/blob/main/examples/StereoDepth/rgb_depth_aligned.cpp

Perfect. Thanks.

nindanaoto committed:
Added rectify and aligned RGB

nindanaoto committed:
change rgbaligned to depth_aligned

nindanaoto commented:
@saching13 
I added the aligned RGB image in rgb_stereo_inertial.cpp. This causes lowering FPS of stereo images, so I kept stereo_inertial.cpp as a separate file.
Along with this modification, I added `rectify` option to publish rectified images instead of raw images.

I'm still not sure about the exact position of IMU. I think this should be calculated from the PCB design.

nindanaoto commented:
I tried colorizing point clouds by using this nodelet with `rgb_stereo_inertial.cpp` but failed. I'm not sure about the reason. This should be left for other PRs or issues since this requires introducing additional dependencies and has nothing to do with handling the IMU data.
http://wiki.ros.org/depth_image_proc

saching13 commented:
> @saching13 I added the aligned RGB image in rgb_stereo_inertial.cpp. This causes lowering FPS of stereo images, so I kept stereo_inertial.cpp as a separate file. Along with this modification, I added `rectify` option to publish rectified images instead of raw images.
> 
> I'm still not sure about the exact position of IMU. I think this should be calculated from the PCB design.

You can just keep it one example and have rgb enabled only when align is set or else you can just publish depth/Disparity along with rectified right. Since that is where it is aligned.

nindanaoto committed:
RGB will be disbled when depth_aligned is false

nindanaoto commented:
I fixed it.  If depth_aligend is false, the RGB image is disabled. Is this OK?

nindanaoto commented:
I found that this implementation causes strange behavior. When I set depth_aligned to true, IMU's data rate becomes around 170Hz but when I set it to false, only around 30 Hz.

nindanaoto committed:
moving definition of queue resolved IMU data rate issue

nindanaoto commented:
Fixed by moving the definition of the queue for the RGB image. I guess that only defining but not using the queue causes some performance issues.

saching13 commented:
> Fixed by moving the definition of the queue for the RGB image. I guess that only defining but not using the queue causes some performance issues.

Yes. because it will use the bandwidth to send the images. 
And also can you make sure to create RGB node and xlink node for that in pipeline only when using ? 
And also remove the rectified images when the image is aligned to rgb. The reason being they don't serve any purpose since depth is in rgb camera frame which would in rectified right camera frame when it is not aligned to RGB.

nindanaoto committed:
Merge branch 'luxonis:noetic-devel' into noetic-devel

FPSychotic commented:
Thanks all of you.

nindanaoto committed:
Seems to be working as expected

nindanaoto committed:
set subpixel on as default

nindanaoto committed:
Merge branch 'luxonis:noetic-devel' into noetic-devel

nindanaoto commented:
@saching13 
I think all issues are fixed. How about this version?

nindanaoto commented:
The rectify option is now off in the default setting because we can now undistort images by [stereo_image_proc](http://wiki.ros.org/stereo_image_proc).

saching13 commented:
Thanks alot. Will check in sometime.

saching13 commented on depthai_examples/src/stereo_inertial_publisher.cpp:
```diff
@@ -0,0 +1,277 @@
+    // XLinkOut
+    xoutLeft->setStreamName("left");
+    xoutRight->setStreamName("right");
+    xoutRectifiedLeft->setStreamName("rectifiedleft");
+    xoutRectifiedRight->setStreamName("rectifiedright");
```
These are not used. You can remove them since you are using xoutLeft and xoutRight. 
Or modify it based on which you are going to use.

saching13 commented on depthai_examples/src/stereo_inertial_publisher.cpp:
```diff
@@ -0,0 +1,277 @@
+    // StereoDepth
+    stereo->initialConfig.setConfidenceThreshold(200);
+    stereo->setRectifyEdgeFillColor(0); // black, to better see the cutout
+    stereo->initialConfig.setLeftRightCheckThreshold(1);
```
set this to 10. We saw better results with 10.

nindanaoto replied:
Which one I should set to 10? ConfidenceThreshold or LeftRightCheckThreshold or both?

saching13 replied:
sorry my bad. `LeftRightCheckThreshold` to 10. And also set `ConfidenceThreshold` 220

nindanaoto replied:
Fixed.

saching13 commented on depthai_examples/src/stereo_inertial_publisher.cpp:
```diff
@@ -0,0 +1,277 @@
+        if(depth_aligned){
+            auto imgQueue = device.getOutputQueue("rgb", 30, false);
+            dai::rosBridge::ImageConverter rgbConverter(deviceName + "_rgb_camera_optical_frame", false);
+            auto rgbCameraInfo = rgbConverter.calibrationToCameraInfo(calibrationHandler, dai::CameraBoardSocket::RGB, 1920, 1080); 
```
when depth is aligned to rgb. CameraInfo of depth will be same as of rgb instead of rectified right

nindanaoto replied:
I've tried to change that but this will cause a mismatch of the resolution between the depth cloud and the camera info on rviz. So, I'm currently keeping to use rightCameraInfo. What is the original intention for changing this camera info?

saching13 replied:
The reason why we need to change is because the intrinsics changes when we match the depth to rgb camera. So we need to use the rgb camera intrincs but rescaled version of it to match to the resolution of the depth. If this is not done point cloud conversion will be wrong.

saching13 replied:
`auto depthCameraInfo = converter.calibrationToCameraInfo(calibrationHandler, dai::CameraBoardSocket::RGB, 1280, 720);` // This should work if the res of depth is 720p

nindanaoto replied:
This may be fixed in the current HEAD but I couldn't see the point cloud. The error is like the following.
`Error updating depth cloud: Depth image has invalid format (only 16 bit and float are supported)!`
This is reproduced by running stereo_node.launch, so I think this is another issue, not related to this commit.

saching13 replied:
You can merge once changing the camera_info for disparity. Thanks a lot for the PR and the changes.

saching13 commented on depthai_examples/src/stereo_inertial_publisher.cpp:
```diff
@@ -0,0 +1,277 @@
+            ros::spin();
+        }
+        else {
+            dai::rosBridge::BridgePublisher<sensor_msgs::Image, dai::ImgFrame> leftPublish(leftQueue,
+                                                                                            pnh, 
+                                                                                            leftPubName,
+                                                                                            std::bind(&dai::rosBridge::ImageConverter::toRosMsg, 
+                                                                                            &converter, 
+                                                                                            std::placeholders::_1, 
+                                                                                            std::placeholders::_2) , 
+                                                                                            30,
+                                                                                            leftCameraInfo,
+                                                                                            "left");
+            dai::rosBridge::BridgePublisher<sensor_msgs::Image, dai::ImgFrame> rightPublish(rightQueue,
+                                                                                            pnh, 
+                                                                                            rightPubName,
+                                                                                            std::bind(&dai::rosBridge::ImageConverter::toRosMsg, 
+                                                                                            &rightconverter, 
+                                                                                            std::placeholders::_1, 
+                                                                                            std::placeholders::_2) , 
+                                                                                            30,
+                                                                                            rightCameraInfo,
+                                                                                            "right");  
```
if we are not publishing left and right when aligned to rgb. We can remove the xlink of these in pipeline init to avoid unused usb bandwidth.

saching13 commented on depthai_examples/src/stereo_inertial_publisher.cpp:
```diff
@@ -0,0 +1,277 @@
+    }
+    else{
+        dai::rosBridge::DisparityConverter dispConverter(deviceName + "_right_camera_optical_frame", 880, 7.5, 20, 2000);
+        dai::rosBridge::BridgePublisher<stereo_msgs::DisparityImage, dai::ImgFrame> dispPublish(stereoQueue,
+                                                                                     pnh, 
+                                                                                     std::string("stereo/disparity"),
+                                                                                     std::bind(&dai::rosBridge::DisparityConverter::toRosMsg, 
+                                                                                     &dispConverter, 
+                                                                                     std::placeholders::_1, 
+                                                                                     std::placeholders::_2) , 
+                                                                                     30,
+                                                                                     rightCameraInfo,
+                                                                                     "stereo");
```
We can publish disparity aligned with rgb also. Doesn't need to be only with depth.

saching13 reviewed (changes requested):
Thanks for all the changes. I added some more suggestions for better USB bandwidth usage.

nindanaoto committed:
fixed parameters and reduce bandwidth cosumption

nindanaoto committed:
Merge branch 'noetic-devel' of https://github.com/nindanaoto/depthai-ros-examples into noetic-devel

nindanaoto committed:
fix default param

nindanaoto committed:
use rgb info for depth when aligned

saching13 commented on depthai_examples/src/stereo_inertial_publisher.cpp:
```diff
@@ -0,0 +1,295 @@
+                                                                                     std::placeholders::_1, 
+                                                                                     std::placeholders::_2) , 
+                                                                                     30,
+                                                                                     rightCameraInfo,
```
use scaled rgb `camera_info` for disparity also when aligned to RGB

nindanaoto committed:
Fix disparityCameraInfo

nindanaoto commented:
Fixed the camera_info for the disparity. Is it OK?

saching13 commented:
Yup. Thanks.
